lib/rucio/daemons/rsedecommissioner/profiles/generic.py

```diff
@@ -0,0 +1,489 @@
+               rse['rse'], limit_check.deleted_rules)
+    elif limit_check.processed_rules != 0:
+        # Nothing was processed in this round but there are rules that we did not
+        # know how to process
+        logger(logging.WARNING,
+               '(%s) %s rules have nontrivial RSE expressions and multiple'
+               ' replicas. We need to move the replicas and/or update the rules'
+               ' before proceeding with decommissioning.',
+               rse['rse'], limit_check.processed_rules)
+
+        set_status(rse['id'], DecommissioningStatus.SUSPENDED)
+    else:
+        # Check for (and potentially resolve) remaining replicas.
+        _generic_finalize(rse, logger=logger)
+
+
+def _generic_initialize(
+    rse: 'dict[str, Any]',
+    status: DecommissioningStatus,
+    *,
+    logger: 'Callable[..., None]' = logging.log,
+) -> None:
+    """Set the RSE availability flags and the decommissioning status.
+
+    When an RSE is initialized for decommissioning, the availability flags are
+    set as follows:
+
+    - `availability_read=True`
+    - `availability_write=False`
+    - `availability_delete=True`
+
+    and the following attribute is set:
+
+    - `greedyDeletion=True`
+
+    :param rse: RSE table entry as a dictionary.
+    :param status: Decommissioning status.
+    :param logger: Logging function.
+    """
+
+    if status != DecommissioningStatus.PROCESSING:
+        return
+
+    # Set the RSE availability and greedyDeletion attributes.
+    logger(logging.INFO,
+           '(%s) Setting RSE availability flags of to read !write delete,'
+           ' greedyDeletion=true.', rse['rse'])
+
+    parameters = {
+        'availability_read': True,
+        'availability_write': False,
+        'availability_delete': True
+    }
+    update_rse(rse['id'], parameters)
+
+    try:
+        add_rse_attribute(rse['id'], 'greedyDeletion', True)
+    except Duplicate:
+        pass
+
+
+class RuleRemovability(Enum):
+    """Rule removability flags."""
+    DONT_REMOVE = 0
+    CAN_REMOVE = 1
+    BEING_REMOVED = 2
+
+
+def _check_removability(
+    rule: models.ReplicationRule,
+    rse: 'dict[str, Any]',
+    *,
+    logger: 'Callable[..., None]' = logging.log,
+) -> RuleRemovability:
+    """Determine if the rule can be removed.
+
+    This function considers a rule removable if its RSE expression is trivial
+    (RSE name itself) or if all locks for the rule is at the given RSE.
+
+    :param rule: Rule to delete.
+    :param rse: RSE table entry as a dictionary.
+    :param logger: Logging function.
+    :returns: Removability of the rule.
+    """
+    # Is this rule already set to be deleted?
+    if (rule.expires_at is not None and rule.expires_at < datetime.utcnow()):
+        logger(logging.DEBUG,
+               '(%s) Rule %s for %s %s:%s is bound for deletion', rse['rse'],
+               rule.id, rule.did_type, rule.scope, rule.name)
```

Comment:
You’re using `rule.did_type` in a couple of places. You should note that it’s an `Enum`, not an `str`. Its use here is wrong, but I’m not sure it’s helpful to have in the logs at all.

Reply:
Dropped all uses of rule.did_type in log messages.